modules/tensor_mechanics/src/materials/ADComputeFiniteShellStrain.C

```diff
@@ -0,0 +1,378 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "ADComputeFiniteShellStrain.h"
+#include "DenseMatrix.h"
+
+#include "libmesh/quadrature.h"
+#include "libmesh/utility.h"
+#include "libmesh/enum_quadrature_type.h"
+#include "libmesh/fe_type.h"
+#include "libmesh/string_to_enum.h"
+#include "libmesh/quadrature_gauss.h"
+
+registerADMooseObject("TensorMechanicsApp", ADComputeFiniteShellStrain);
+
+defineADValidParams(ADComputeFiniteShellStrain,
+                    ADComputeIncrementalShellStrain,
+                    params.addClassDescription("Compute a large strain increment for the shell."););
+
+template <ComputeStage compute_stage>
+ADComputeFiniteShellStrain<compute_stage>::ADComputeFiniteShellStrain(
+    const InputParameters & parameters)
+  : ADComputeIncrementalShellStrain<compute_stage>(parameters), _BNL()
+{
+  _BNL.resize(_t_points.size());
+
+  for (unsigned int i = 0; i < _t_points.size(); ++i)
+    _BNL[i] = &declareADProperty<DenseMatrix<Real>>("BNL_matrix_t_points_" + std::to_string(i));
+}
+
+template <ComputeStage compute_stage>
+void
+ADComputeFiniteShellStrain<compute_stage>::initQpStatefulProperties()
+{
+  computeGMatrix();
+
+  ADDenseMatrix b(5, 20);
+  for (unsigned int t = 0; t < _t_points.size(); ++t)
+    (*_BNL[t])[_qp] = b;
+}
+
+template <ComputeStage compute_stage>
+void
+ADComputeFiniteShellStrain<compute_stage>::computeProperties()
+{
+  // quadrature points in isoparametric space
+  _2d_points = _qrule->get_points(); // would be in 2D
+
+  unsigned int dim = _current_elem->dim();
+  if ((dim != 2))
+    mooseError("Shell element is implemented only for 2D Linear elements");
```

Comment:
Linear -> linear

Reply:
Can't we do this check earlier on so we're not doing it every material evaluation?  Also, probably need to check for the order of the element and for the fact that it's restricted to 4-noded elements as well (it doesn't work for triangles, right?)

Reply:
I have moved this check to initQpStatefulProperties of ADComputeIncrementalShellStrain. 

I have also added a test to check if that the element is first order (in the constructor) and to check if the element has 4 nodes (in initQpStatefulProperties).